Added support for valexpire column for set tabletype only.

deem0n opened the pull request:
Hello, 

Thanks for your work, TTL handling is very useful. I just extended your code to support expiry for the individual set members.
Probably it will be more fixes soon, after we run rw_redis_fdw with real data.

valexpire column allows to set expiration for the individual member of the set. It works similar to the expiry column.
It uses EXPIREMEMBER command which is available only in KeyDB. It is not available in Redis as of 2020.
https://docs.keydb.dev/docs/commands/#expiremember

Also fixed bug with error handling, so we really print error message instead of crashing in many cases.

Some more work is required, especially documentation.

deem0n committed:
Added support for valexpire column for set tabletype only.
valexpire column allows to set expiration for the individual member of the set. It works similar to the expiry column.
It uses EXPIREMEMBER command which is available only in KeyDB. It is not available in Redis as of 2020.
https://docs.keydb.dev/docs/commands/#expiremember

Also fixed bug with error handling, so we really print error message instead of crashing in many cases.

Some more work is required, especially documentation.

deem0n committed:
bump version to 1.0.1

deem0n committed:
remove warning and fix build for 1.0.1

l-d-x commented:
Thanks for the patch. So this is specifically for KeyDB and not Redis? Also, why VALEXPIRE instead of its EXPIREMEMBER name?

On the outset, wouldn't it be more appropriate to have that done inside the TTL table with an optional `member` column?

I'll try to review in detail as soon as I can.

deem0n commented:
Hi,

Yes, it works only with KeyDB. Redis has no EXPIREMEMBER command.

I used VALEXPIRE because I hope that someday Redis or KeyDB or something like Aerospike will support HKEY subkey expiration. So VALEXPIRE or EXPIREVAL is more generic name comparing to EXPIREMEMBER. Now I think EXPIREVAL is a better name, I hope we can use it!

The idea with TTL table can be implemented, but that way you need two operations to control EXPIRE for the set member. 1) you need to insert into 'set' table, 2) you need to assign expiration with the TTL table.

Currently, you can do it with one insert into 'set' table. Even FDW `DEFAULT` works, which is very handy:

```sql
CREATE FOREIGN TABLE portal.session_details (
    counter TEXT OPTIONS (redis 'key'), 
    member  TEXT,
    expire  INT OPTIONS (redis 'valexpire') DEFAULT 3600
    )
  SERVER redis_server
  OPTIONS (database '0', tabletype 'set');
```

So if you do `INSERT INTO portal.session_details VALUES ('a','b')` member b will be assigned 1 hour expiration time automatically.

Also, both `valexpire` and `expiry` can be used at the same time and working fine for 'set' table.

So I think we can add `valexpire` for TTL table as well, but I prefer to keep it for 'set' table as well.

deem0n commented:
Also, it might be interesting to you, I tried to do the same idea with the original redis_fdw: https://github.com/luxms/redis_fdw/commit/bbed188f6d4b37b529b382ab1b16dd715845ac8a https://github.com/luxms/redis_fdw/blob/REL_11_STABLE/redis_fdw.c  (Only this branch)

That way you set default expiration time with table OPTIONS: `OPTIONS (database '0', tabletype 'set', keyexpire '3600')`.

The problem with original redis_fdw for me is that there is no way you can get collection size without iterating over collection. In other words there is no `count(*)` aggregate push down and it is very hard for me to implement it. So 'len' table from your project helped me.

deem0n committed:
Renamed valexpire to valttl

l-d-x commented:
I'll spend some time over the weekend to a look at your diffs.

Postgres v10+ supports aggregate pushdown for COUNT(*), but I haven't had time to study it yet and would likely need an example before any work can be done here so we can leverage it.

l-d-x commented on Makefile:
```diff
@@ -3,14 +3,14 @@ MODULE_big = redis_fdw
 
 SHLIB_LINK += -lhiredis
 
-PG_CPPFLAGS+= -DWRITE_API -g
```
The -g switch is to help with coredump debugging, and doesn't hurt the performance of the compiled code, so there's no need to move it in the DEBUG section which is more akin to code level debugging and not compiler settings.

l-d-x commented on redis_fdw.c:
```diff
@@ -1437,6 +1450,15 @@ get_psql_columns(Oid foreigntableid, struct redis_fdw_ctx *rctx)
+			/* We can expire individual values only for set and only with KeyDB */
+			if (rctx->table_type == PG_REDIS_SET) {
+				verify_pgtable_coltype(R_INT, att_tuple->atttypid,
+									colname, tablename);
```
Can you convert the second line indentation of multiline code blocks to use tabs to the same indentation level and spaces after, ie tab,tab,tab,tab,spaces....."colname". I understand the indentation of pgsql is tabspaces=4, but it helps readability in browsers that show them as 8, so using spaces after tabs will keep the alignment consistent for any mode.

l-d-x commented on redis_fdw.c:
```diff
@@ -2693,6 +2728,24 @@ redisIterateForeignScan(ForeignScanState *node)
 		}
 
+		/* prefetch valttl if valttl column exists */
+		if ((rctx->rtable.valttl > 0) && (rctx->table_type == PG_REDIS_SET) && (rctx->where_flags & PARAM_MEMBER)) {
```
`if` -> `else if`
and add newline before `(rctx->where_flags & PARAM_MEMBER)` to split line up.

optional: assignment of rtable.valttl was already done and only set if table_type ==  PG_REDIS_SET in Line 1458, so the test for (rctx->table_type == PG_REDIS_SET) is redundant here, but harmless.

l-d-x commented on redis_fdw.c:
```diff
@@ -3146,11 +3202,22 @@ redisIterateForeignScan(ForeignScanState *node)
 		case VAR_MEMBERS:
 			break;
 		case VAR_EXPIRY:
+		    DEBUG((DEBUG_LEVEL, "VAR_EXPIRY CASE"));
```
remove this debug message; I think you had it here to verify your test case.

l-d-x commented on redis_fdw.c:
```diff
@@ -3146,11 +3202,22 @@ redisIterateForeignScan(ForeignScanState *node)
+			}
+			break;
+		case VAR_VALTTL:
+		    DEBUG((DEBUG_LEVEL, "VAR_VALTTL CASE"));
```
as above, remove this debug message.

l-d-x commented on redis_fdw.c:
```diff
@@ -3003,7 +3055,10 @@ redisIterateForeignScan(ForeignScanState *node)
 	case REDIS_SISMEMBER:
 		redis_get_reply(reply, &s_value, &i_value, &nil_value);
-		if (!nil_value)
+		DEBUG((DEBUG_LEVEL, "  GOT REDIS_SISMEMBER reply"));
```
The original code has a bug (I've just pushed a fix), so I'm glad you brought up something here. It just happens that if a member matches, no value was being returned which is a bug. And s_value should be set when the member exists. The correction is:

```
-               if (!nil_value)
-                       nil_value = (i_value != 0);
-               else
-                       s_value = rctx->where_conds.s_value;
+
+               /* If no match, then set nil_value to true */
+               nil_value = (i_value == 0);
+               s_value = rctx->where_conds.s_value;
```

l-d-x reviewed (commented):
Some minor cleanup. I'll integrate once you've done it and verified testing.

deem0n committed:
2020-12-03 Version 1.0.2

- Allow updates for valttl field.
- Fix crash in some rare cases.
- Fix print format for int64_t

deem0n committed:
fix printf

deem0n committed:
fix printf

deem0n committed:
remove warnings on printf

deem0n committed:
Merge branch 'master' of https://github.com/nahanni/rw_redis_fdw

deem0n committed:
Makefile

deem0n committed:
Fixed bug with Redis format string

deem0n committed:
bump version to 1.0.3

deem0n committed:
Added *.sql to Makefile

deem0n committed:
Added .spec for building RPM

deem0n committed:
removed all %jd in favor of %lld

deem0n committed:
safe conversion using %lld and (long long)

deem0n committed:
using PRId64 for int64_t

deem0n committed:
put back %lld for redis struct integer (long long)

achistyako committed:
Update redis_fdw.c

deem0n committed:
Merge pull request #3 from achistyako/master

Support for PostgreSQL 14

achistyako committed:
Update redis_fdw.c

deem0n committed:
Merge pull request #4 from achistyako/master

fix new callback signature for Pg 14

deem0n committed:
Update README.md

achistyako committed:
Update redis_fdw.c

achistyako committed:
Merge pull request #5 from achistyako/master

Update redis_fdw.c

achistyako committed:
Update redis_fdw.c

achistyako committed:
Merge pull request #6 from achistyako/master

Update redis_fdw.c

achistyako committed:
Create  redis_fdw--1.0.5.sql

achistyako committed:
Create redis_fdw--1.0.4--1.0.5.sql

achistyako committed:
Update redis_fdw.control

achistyako committed:
Update rw_redis_fdw.spec

achistyako committed:
Update rw_redis_fdw.spec

achistyako committed:
Update rw_redis_fdw.spec

achistyako committed:
Merge pull request #7 from achistyako/master

version change

achistyako committed:
Update Makefile

achistyako committed:
Merge pull request #8 from achistyako/master

Update Makefile

achistyako committed:
Update Makefile

achistyako committed:
Merge pull request #9 from achistyako/master

Update Makefile wildcard

Commit:
add PostgeSQL 15 initial compatibility

Commit:
update readme

Commit:
bump extension version

Commit:
update changelogs and add missing entries

Commit:
fix formatting

deem0n committed:
Merge pull request #10 from petrov9268/master

PostgreSQL 15 initial support

Commit:
redis_fdw.c: save key with internal name, not real, fixes UPDATE/DELETE

When changes for pg14 were applied, key was saved with real column name,
which lead to "key not provided" error when there was function call in WHERE
statement.

See https://github.com/postgres/postgres/commit/86dc90056dfdbd9d1b891718d2e5614e3e432f35
for explanation of changes between pg13 and pg14 SDK that lead to these
changes in the first place.

Commit:
redis_fdw.sql: relabel key and use format(TEXT, TEXT) in WHERE

This might help to catch similar errors in the future.

Commit:
add expected output file

petrov9268 committed:
add update sql files

petrov9268 committed:
update changelog, bump version

deem0n committed:
Merge pull request #11 from petrov9268/master

fix UPDATE/DELETE operations with function calls

petrov9268 committed:
redis_fdw.c: add T_OpExpr to condition

Similar to https://github.com/nahanni/rw_redis_fdw/issues/14. On sixth run,
Postgres may switch the subquery in a function with a where clause from a
T_ConstExpr to T_OpExpr, which will lead to 'key is NULL' error without
any context.

petrov9268 committed:
redis_fdw.c: remove trailing spaces

petrov9268 committed:
expected: remove localized lines

petrov9268 committed:
redis_fdw.sql: add order by for consistent output

petrov9268 committed:
add functions that caused errors, remove localized strings, change expected

petrov9268 committed:
redis_fdw.c: fix last warning with -Wimplicit-fallthrough=3

petrov9268 committed:
move all upgrade files to subdirectory, bump version to 1.1

and111016 committed:
Update Makefile (#13)

* Update Makefile

Fix for builds

* Create redis_fdw--1.0.8.sql

* Create redis_fdw--1.0.7--1.0.8.sql

* Update CHANGES

and111016 committed:
And111016 patch 1 (#14)

* Update Makefile

Fix for builds

* Create redis_fdw--1.0.8.sql

* Create redis_fdw--1.0.7--1.0.8.sql

* Update CHANGES

* Fix redis_fdw.control

deem0n committed:
Merge branch 'master' into master

deem0n committed:
Merge pull request #12 from petrov9268/master

fix Key is NULL error

deem0n committed:
Revert "fix Key is NULL error"

deem0n committed:
Merge pull request #15 from luxms/revert-12-master

Revert "fix Key is NULL error"

petrov9268 committed:
redis_fdw.c: add T_OpExpr to condition

Similar to nahanni#14. On sixth run,
Postgres may switch the subquery in a function with a where clause from a
T_ConstExpr to T_OpExpr, which will lead to 'key is NULL' error without
any context.

petrov9268 committed:
redis_fdw.c: remove trailing spaces

petrov9268 committed:
redis_fdw.c: fix last warning with -Wimplicit-fallthrough=3

petrov9268 committed:
expected: remove localized lines

petrov9268 committed:
redis_fdw.sql: add order by for consistent output

petrov9268 committed:
redis_fdw.sql:  add functions that cause errors, add expected output

petrov9268 committed:
fix update 1.0.8

petrov9268 committed:
bump version to 1.0.9

deem0n committed:
Merge pull request #16 from petrov9268/key-is-null-error

Key is null error

petrov9268 committed:
redis_fdw.c: remove incorrect Assert, fix implicit fallthrough

* this Assert seems to be incorrect, most of the time param field contains
  NULL, this leads to immediate crash on any INSERT attempt with
  --enable-cassert
* for some reason compiler never complained about implicit fallthrough, but
  it seems like an error, since parameter is not actually skipped
* fix formatting

petrov9268 committed:
redis_fdw.c: restore key and where flags on each redisIterateForeignScan call

IterateForeignScan callback is called in a short-lived memory context that
will be reset between invocations according to documentation:
https://www.postgresql.org/docs/current/fdw-callbacks.html#FDW-CALLBACKS-SCAN

for now restore key value on each call, not only on first run

petrov9268 committed:
Makefile: treat all warnings as errors, disable optimization in debug

petrov9268 committed:
redis_fdw.c: call to elog must be last, copy string for error message

previous crash fix was incorrect: call to elog MUST be last in
ERR_CLEANUP, otherwise reply and connection allocated by hiredis might
leak from my understanding, since elog with severity level above ERROR
never returns to caller

to fix initial issue, copy error string and pass copy to ERR_CLEANUP,
which should print error and properly release hiredis at the same time
resources

petrov9268 committed:
redis_fdw.c: use same format for fall-through

petrov9268 committed:
redis_fdw.c: address shaky error checking

* also do not check isnil in redis_get_reply

petrov9268 committed:
fix another two error checks, free reply in function

petrov9268 committed:
bump version

petrov9268 committed:
redis_fdw.c: fix a few more error check mistakes

petrov9268 committed:
redis_fdw.c: rowcount is unsigned, and can not be less than 0

deem0n committed:
Merge pull request #19 from petrov9268/cassert-fix

fix issues related to --enable-cassert, address memory issues on redis error checks

petrov9268 committed:
redis_fdw.c: fix error

older compilers complained with:
redis_fdw.c:3053:5: error: a label can only be part of a statement and a
declaration is not a statement

apparently, it is not allowed to declare anything after label, and apparently
default is considered a label?

deem0n committed:
Merge pull request #20 from petrov9268/fix-label-can-only-be-part-of-a-statement

redis_fdw.c: fix error

petrov9268 committed:
redisAddForeignUpdateTargets: param_flags never used

deem0n committed:
Merge pull request #21 from petrov9268/remove-unused-var

redisAddForeignUpdateTargets: param_flags never used

petrov9268 committed:
redis_fdw.c: compile for pgsql 16, need to test for regressions

petrov9268 committed:
redis_fdw.c: compile for pgsql 17, need to test for regressions

petrov9268 committed:
redis_fdw.c: fix PG_VERSION_NUM

petrov9268 committed:
redis_fdw.c: conditionally include inherit.h, call function for pg>=15.2

petrov9268 committed:
redis_fdw.c: fix formatting

petrov9268 committed:
add ability to run make installcheck to run test files

deem0n committed:
Merge pull request #22 from petrov9268/pgsql-16-17

pgsql 16 and 17 support

petrov9268 committed:
makefile: generate redundant sql files

petrov9268 committed:
cleanup:

* remove outdated spec file and RPM section from README.md
* bump version

deem0n committed:
Merge pull request #24 from petrov9268/release-1.0.11

Release 1.0.11
